In [14]:

import pandas as pd
import numpy as np

# read the csv dataset for diamonds

df = pd.read_csv("diamonds.csv")
df.head()


Unnamed: 0  carat      cut color clarity  depth  table  price     x     y  \
0           1   0.23    Ideal     E     SI2   61.5   55.0    326  3.95  3.98   
1           2   0.21  Premium     E     SI1   59.8   61.0    326  3.89  3.84   
2           3   0.23     Good     E     VS1   56.9   65.0    327  4.05  4.07   
3           4   0.29  Premium     I     VS2   62.4   58.0    334  4.20  4.23   
4           5   0.31     Good     J     SI2   63.3   58.0    335  4.34  4.35   

      z  
0  2.43  
1  2.31  
2  2.31  
3  2.63  
4  2.75

In [15]:
#Number of records in dataset
df.tail()

Unnamed: 0  carat        cut color clarity  depth  table  price     x  \
53935       53936   0.72      Ideal     D     SI1   60.8   57.0   2757  5.75   
53936       53937   0.72       Good     D     SI1   63.1   55.0   2757  5.69   
53937       53938   0.70  Very Good     D     SI1   62.8   60.0   2757  5.66   
53938       53939   0.86    Premium     H     SI2   61.0   58.0   2757  6.15   
53939       53940   0.75      Ideal     D     SI2   62.2   55.0   2757  5.83   

          y     z  
53935  5.76  3.50  
53936  5.75  3.61  
53937  5.68  3.56  
53938  6.12  3.74  
53939  5.87  3.64

In [16]:
#Clean the data:
#Create list of true/false values
dupe = df.duplicated() 
#Shwo rows where duplicate is true
df[dupe] 

Empty DataFrame
Columns: [Unnamed: 0, carat, cut, color, clarity, depth, table, price, x, y, z]
Index: []

In [17]:
#show data types in the dataset
df.dtypes


Unnamed: 0      int64
carat         float64
cut            object
color          object
clarity        object
depth         float64
table         float64
price           int64
x             float64
y             float64
z             float64
dtype: object

In [18]:
#Count rows including NA values
len(df)

53940

In [19]:
#number of non-NA values
df.count()


Unnamed: 0    53940
carat         53940
cut           53940
color         53940
clarity       53940
depth         53940
table         53940
price         53940
x             53940
y             53940
z             53940
dtype: int64

In [20]:
#drop rows with any missing data
df_no_missing = df.dropna()
df_no_missing

Unnamed: 0  carat        cut color clarity  depth  table  price     x  \
0               1   0.23      Ideal     E     SI2   61.5   55.0    326  3.95   
1               2   0.21    Premium     E     SI1   59.8   61.0    326  3.89   
2               3   0.23       Good     E     VS1   56.9   65.0    327  4.05   
3               4   0.29    Premium     I     VS2   62.4   58.0    334  4.20   
4               5   0.31       Good     J     SI2   63.3   58.0    335  4.34   
5               6   0.24  Very Good     J    VVS2   62.8   57.0    336  3.94   
6               7   0.24  Very Good     I    VVS1   62.3   57.0    336  3.95   
7               8   0.26  Very Good     H     SI1   61.9   55.0    337  4.07   
8               9   0.22       Fair     E     VS2   65.1   61.0    337  3.87   
9              10   0.23  Very Good     H     VS1   59.4   61.0    338  4.00   
10             11   0.30       Good     J     SI1   64.0   55.0    339  4.25   
11             12   0.23      Ideal     J     VS1   62.8   56.0    340  3.93   
12             13   0.22    Premium     F     SI1   60.4   61.0    342  3.88   
13             14   0.31      Ideal     J     SI2   62.2   54.0    344  4.35   
14             15   0.20    Premium     E     SI2   60.2   62.0    345  3.79   
15             16   0.32    Premium     E      I1   60.9   58.0    345  4.38   
16             17   0.30      Ideal     I     SI2   62.0   54.0    348  4.31   
17             18   0.30       Good     J     SI1   63.4   54.0    351  4.23   
18             19   0.30       Good     J     SI1   63.8   56.0    351  4.23   
19             20   0.30  Very Good     J     SI1   62.7   59.0    351  4.21   
20             21   0.30       Good     I     SI2   63.3   56.0    351  4.26   
21             22   0.23  Very Good     E     VS2   63.8   55.0    352  3.85   
22             23   0.23  Very Good     H     VS1   61.0   57.0    353  3.94   
23             24   0.31  Very Good     J     SI1   59.4   62.0    353  4.39   
24             25   0.31  Very Good     J     SI1   58.1   62.0    353  4.44   
25             26   0.23  Very Good     G    VVS2   60.4   58.0    354  3.97   
26             27   0.24    Premium     I     VS1   62.5   57.0    355  3.97   
27             28   0.30  Very Good     J     VS2   62.2   57.0    357  4.28   
28             29   0.23  Very Good     D     VS2   60.5   61.0    357  3.96   
29             30   0.23  Very Good     F     VS1   60.9   57.0    357  3.96   
...           ...    ...        ...   ...     ...    ...    ...    ...   ...   
53910       53911   0.70    Premium     E     SI1   60.5   58.0   2753  5.74   
53911       53912   0.57    Premium     E      IF   59.8   60.0   2753  5.43   
53912       53913   0.61    Premium     F    VVS1   61.8   59.0   2753  5.48   
53913       53914   0.80       Good     G     VS2   64.2   58.0   2753  5.84   
53914       53915   0.84       Good     I     VS1   63.7   59.0   2753  5.94   
53915       53916   0.77      Ideal     E     SI2   62.1   56.0   2753  5.84   
53916       53917   0.74       Good     D     SI1   63.1   59.0   2753  5.71   
53917       53918   0.90  Very Good     J     SI1   63.2   60.0   2753  6.12   
53918       53919   0.76    Premium     I     VS1   59.3   62.0   2753  5.93   
53919       53920   0.76      Ideal     I    VVS1   62.2   55.0   2753  5.89   
53920       53921   0.70  Very Good     E     VS2   62.4   60.0   2755  5.57   
53921       53922   0.70  Very Good     E     VS2   62.8   60.0   2755  5.59   
53922       53923   0.70  Very Good     D     VS1   63.1   59.0   2755  5.67   
53923       53924   0.73      Ideal     I     VS2   61.3   56.0   2756  5.80   
53924       53925   0.73      Ideal     I     VS2   61.6   55.0   2756  5.82   
53925       53926   0.79      Ideal     I     SI1   61.6   56.0   2756  5.95   
53926       53927   0.71      Ideal     E     SI1   61.9   56.0   2756  5.71   
53927       53928   0.79       Good     F     SI1   58.1   59.0   2756  6.06   
53928  

In [21]:
#Run descriptive statistics 
df.describe()

Unnamed: 0         carat         depth         table         price  \
count  53940.000000  53940.000000  53940.000000  53940.000000  53940.000000   
mean   26970.500000      0.797940     61.749405     57.457184   3932.799722   
std    15571.281097      0.474011      1.432621      2.234491   3989.439738   
min        1.000000      0.200000     43.000000     43.000000    326.000000   
25%    13485.750000      0.400000     61.000000     56.000000    950.000000   
50%    26970.500000      0.700000     61.800000     57.000000   2401.000000   
75%    40455.250000      1.040000     62.500000     59.000000   5324.250000   
max    53940.000000      5.010000     79.000000     95.000000  18823.000000   

                  x             y             z  
count  53940.000000  53940.000000  53940.000000  
mean       5.731157      5.734526      3.538734  
std        1.121761      1.142135      0.705699  
min        0.000000      0.000000      0.000000  
25%        4.710000      4.720000      2.910000  
50%        5.700000      5.710000      3.530000  
75%        6.540000      6.540000      4.040000  
max       10.740000     58.900000     31.800000

In [27]:
import statsmodels.formula.api as smf

In [23]:
#Run a correlation on the dataset.
df.corr()


Unnamed: 0     carat     depth     table     price         x  \
Unnamed: 0    1.000000 -0.377983 -0.034800 -0.100830 -0.306873 -0.405440   
carat        -0.377983  1.000000  0.028224  0.181618  0.921591  0.975094   
depth        -0.034800  0.028224  1.000000 -0.295779 -0.010647 -0.025289   
table        -0.100830  0.181618 -0.295779  1.000000  0.127134  0.195344   
price        -0.306873  0.921591 -0.010647  0.127134  1.000000  0.884435   
x            -0.405440  0.975094 -0.025289  0.195344  0.884435  1.000000   
y            -0.395843  0.951722 -0.029341  0.183760  0.865421  0.974701   
z            -0.399208  0.953387  0.094924  0.150929  0.861249  0.970772   

                   y         z  
Unnamed: 0 -0.395843 -0.399208  
carat       0.951722  0.953387  
depth      -0.029341  0.094924  
table       0.183760  0.150929  
price       0.865421  0.861249  
x           0.974701  0.970772  
y           1.000000  0.952006  
z           0.952006  1.000000

In [24]:
#1. Run a regression.   How does price relate to carat, x, y and z? 
result = smf.ols('price ~ carat + x + y + z', data=df).fit()

In [25]:
result.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.854
Model:                            OLS   Adj. R-squared:                  0.854
Method:                 Least Squares   F-statistic:                 7.892e+04
Date:                Thu, 07 Mar 2019   Prob (F-statistic):               0.00
Time:                        15:52:41   Log-Likelihood:            -4.7187e+05
No. Observations:               53940   AIC:                         9.437e+05
Df Residuals:                   53935   BIC:                         9.438e+05
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept   1921.1740    104.373     18.407      0.000    1716.601    2125.747
carat       1.023e+04     62.937    162.607      0.000    1.01e+04    1.04e+04
x           -884.2091     40.470    -21.848      0.000    -963.532    -804.887
y            166.0384     25.858      6.421      0.000     115.356     216.721
z           -576.2035     39.282    -14.668      0.000    -653.197    -499.210
==============================================================================
Omnibus:                    14400.324   Durbin-Watson:                   1.198
Prob(Omnibus):                  0.000   Jarque-Bera (JB):           336485.128
Skew:                           0.743   Prob(JB):                         0.00
Kurtosis:                      15.145   Cond. No.                         171.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

Result: adj.R-squared is high (0.854) so this model fits; intercept, carat, x and y are statistically significant variables; p-value (0.000) says this is a statistically significant result.  
    

In [26]:
#Remove y from correlation - not strong relationship
result = smf.ols('price ~ carat + x + z', data=df).fit()
result.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.854
Model:                            OLS   Adj. R-squared:                  0.854
Method:                 Least Squares   F-statistic:                 1.051e+05
Date:                Thu, 07 Mar 2019   Prob (F-statistic):               0.00
Time:                        15:52:48   Log-Likelihood:            -4.7189e+05
No. Observations:               53940   AIC:                         9.438e+05
Df Residuals:                   53936   BIC:                         9.438e+05
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept   1935.8588    104.387     18.545      0.000    1731.259    2140.459
carat       1.024e+04     62.955    162.639      0.000    1.01e+04    1.04e+04
x           -737.7802     33.446    -22.059      0.000    -803.335    -672.225
z           -549.5797     39.077    -14.064      0.000    -626.172    -472.988
==============================================================================
Omnibus:                    14353.572   Durbin-Watson:                   1.198
Prob(Omnibus):                  0.000   Jarque-Bera (JB):           334529.363
Skew:                           0.740   Prob(JB):                         0.00
Kurtosis:                      15.110   Cond. No.                         131.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

Result: Adj.R-squared remains the same; intercept increases while x and z decrease; p-value is same. Not much change. Something else is influencing the model.

In [13]:
#set coefficient to 0
result = smf.ols(formula='price ~ carat + x + y + z - 1', data=df).fit()
result.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.926
Model:                            OLS   Adj. R-squared:                  0.926
Method:                 Least Squares   F-statistic:                 1.676e+05
Date:                Wed, 06 Mar 2019   Prob (F-statistic):               0.00
Time:                        14:25:42   Log-Likelihood:            -4.7204e+05
No. Observations:               53940   AIC:                         9.441e+05
Df Residuals:                   53936   BIC:                         9.441e+05
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
carat       9139.4524     20.693    441.660      0.000    9098.893    9180.012
x           -467.1759     33.639    -13.888      0.000    -533.109    -401.243
y            176.4676     25.933      6.805      0.000     125.639     227.297
z           -480.9702     39.062    -12.313      0.000    -557.532    -404.409
==============================================================================
Omnibus:                    14184.664   Durbin-Watson:                   1.111
Prob(Omnibus):                  0.000   Jarque-Bera (JB):           236479.997
Skew:                           0.827   Prob(JB):                         0.00
Kurtosis:                      13.123   Cond. No.                         65.2
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

Result: Adj.r-squared increases(0.926); p-value remains the same; carat is the only IV that increases. 
I conclude that it is the only factor that influences price.   

In [14]:
#2.  How does price relate to carat, depth and table? 
result = smf.ols('price ~ carat + depth + table', data=df).fit()

In [15]:
result.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.854
Model:                            OLS   Adj. R-squared:                  0.854
Method:                 Least Squares   F-statistic:                 1.049e+05
Date:                Wed, 06 Mar 2019   Prob (F-statistic):               0.00
Time:                        14:32:56   Log-Likelihood:            -4.7194e+05
No. Observations:               53940   AIC:                         9.439e+05
Df Residuals:                   53936   BIC:                         9.439e+05
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     1.3e+04    390.918     33.264      0.000    1.22e+04    1.38e+04
carat       7858.7705     14.151    555.356      0.000    7831.035    7886.506
depth       -151.2363      4.820    -31.378      0.000    -160.683    -141.789
table       -104.4728      3.141    -33.259      0.000    -110.630     -98.316
==============================================================================
Omnibus:                    14107.737   Durbin-Watson:                   1.015
Prob(Omnibus):                  0.000   Jarque-Bera (JB):           154392.067
Skew:                           0.945   Prob(JB):                         0.00
Kurtosis:                      11.070   Cond. No.                     5.02e+03
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 5.02e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

Result: adj.R-squared is high (0.854) so this model fits; intercept doesn't make sense; p-value is optimal (0.000); carat is far removed from depth and table.   

In [16]:
#remove carat from regression
result = smf.ols('price ~ depth + table', data=df).fit()
result.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.017
Model:                            OLS   Adj. R-squared:                  0.017
Method:                 Least Squares   F-statistic:                     465.3
Date:                Wed, 06 Mar 2019   Prob (F-statistic):          4.62e-201
Time:                        14:38:40   Log-Likelihood:            -5.2331e+05
No. Observations:               53940   AIC:                         1.047e+06
Df Residuals:                   53937   BIC:                         1.047e+06
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept  -1.508e+04   1004.719    -15.014      0.000   -1.71e+04   -1.31e+04
depth         82.2616     12.445      6.610      0.000      57.869     106.654
table        242.5835      7.979     30.402      0.000     226.944     258.223
==============================================================================
Omnibus:                    15207.938   Durbin-Watson:                   0.041
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            34864.819
Skew:                           1.622   Prob(JB):                         0.00
Kurtosis:                       5.234   Cond. No.                     4.98e+03
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 4.98e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

Results: Adj.R-square plunged so this is not a good model.  Let's try with different IVs. 

In [17]:
#remove depth from regression
result = smf.ols('price ~ carat + table', data=df).fit()
result.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.851
Model:                            OLS   Adj. R-squared:                  0.851
Method:                 Least Squares   F-statistic:                 1.540e+05
Date:                Wed, 06 Mar 2019   Prob (F-statistic):               0.00
Time:                        14:42:06   Log-Likelihood:            -4.7243e+05
No. Observations:               53940   AIC:                         9.449e+05
Df Residuals:                   53937   BIC:                         9.449e+05
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept   1961.9923    171.811     11.419      0.000    1625.241    2298.744
carat       7820.0379     14.225    549.744      0.000    7792.157    7847.919
table        -74.3007      3.018    -24.623      0.000     -80.215     -68.386
==============================================================================
Omnibus:                    13940.362   Durbin-Watson:                   1.003
Prob(Omnibus):                  0.000   Jarque-Bera (JB):           159408.119
Skew:                           0.918   Prob(JB):                         0.00
Kurtosis:                      11.219   Cond. No.                     1.49e+03
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.49e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

Results:  Adj.R-square is back up but not as high as the first model; p-value is same; table is lower. 
    Will use the first model in next step.    

In [18]:
#set coefficient to 0
result = smf.ols(formula='price ~ carat + depth + table - 1', data=df).fit()
result.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.924
Model:                            OLS   Adj. R-squared:                  0.924
Method:                 Least Squares   F-statistic:                 2.194e+05
Date:                Wed, 06 Mar 2019   Prob (F-statistic):               0.00
Time:                        14:46:47   Log-Likelihood:            -4.7249e+05
No. Observations:               53940   AIC:                         9.450e+05
Df Residuals:                   53937   BIC:                         9.450e+05
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
carat       7797.8694     14.175    550.114      0.000    7770.086    7825.653
depth         -6.9151      2.121     -3.261      0.001     -11.072      -2.758
table        -32.4782      2.300    -14.122      0.000     -36.986     -27.970
==============================================================================
Omnibus:                    13942.521   Durbin-Watson:                   0.994
Prob(Omnibus):                  0.000   Jarque-Bera (JB):           157317.193
Skew:                           0.922   Prob(JB):                         0.00
Kurtosis:                      11.161   Cond. No.                         180.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

Results: Adj-R-squared increased so model is a good fit; p-value is same; carat is the only IV that increased.
Conclusion - only carat size influences price.  

In [28]:
Location = "datasets/tamiami.csv"

df = pd.read_csv(Location)
df.head()

Cart Location  Hot Dog Sales  Employees in Nearby Office Buildings  \
0              1            100                                  1600   
1              2             80                                  1200   
2              3            450                                  2800   
3              4            580                                  4300   
4              5            100                                  1400   

   Num of Nearby Restaurants  Num of Other Food Carts Nearby  Price  
0                          8                              12   4.16  
1                          6                              13   4.63  
2                         19                               6   0.50  
3                         19                               2   0.47  
4                          6                              13   4.24

In [29]:
columns = ['location', 'sales', 'employees', 'restaurants', 'foodcarts', 'price']

#change column names for readability
df.columns = columns
df.head()

location  sales  employees  restaurants  foodcarts  price
0         1    100       1600            8         12   4.16
1         2     80       1200            6         13   4.63
2         3    450       2800           19          6   0.50
3         4    580       4300           19          2   0.47
4         5    100       1400            6         13   4.24

In [30]:
#Number of records in dataset
df.tail()

location  sales  employees  restaurants  foodcarts  price
25        26     60        900            5         12   4.64
26        27    200       1700           11         12   2.72
27        28    280       1900           10          8   1.79
28        29    500       3000           15          4   0.48
29        30    460       2900           17          5   0.53

In [31]:
#show data types in the dataset
df.dtypes


location         int64
sales            int64
employees        int64
restaurants      int64
foodcarts        int64
price          float64
dtype: object

In [32]:
#Clean the data:
#Create list of true/false values
dupe = df.duplicated() 
#Shwo rows where duplicate is true
df[dupe] 

Empty DataFrame
Columns: [location, sales, employees, restaurants, foodcarts, price]
Index: []

In [33]:
#drop rows with any missing data
df_no_missing = df.dropna()
df_no_missing

location  sales  employees  restaurants  foodcarts  price
0          1    100       1600            8         12   4.16
1          2     80       1200            6         13   4.63
2          3    450       2800           19          6   0.50
3          4    580       4300           19          2   0.47
4          5    100       1400            6         13   4.24
5          6     50        700            4         11   4.21
6          7    120       1600            9         13   3.53
7          8    560       3500           18          2   0.45
8          9    250       1900           15          9   2.10
9         10    290       2400           13          9   1.59
10        11    520       3100           16          3   0.47
11        12    480       2700           18          4   0.53
12        13    510       3200           17          3   0.55
13        14     90       1600            8         16   4.54
14        15    110       1500            7         13   3.60
15        16    210       1600           10         11   2.81
16        17     40        600            2         14   4.99
17        18    180       1800            7         14   2.82
18        19     90       1500            9         14   4.21
19        20     80       1500            7         14   4.29
20        21    190       1900            8         13   3.13
21        22    240       1500           16         16   2.34
22        23    300       2500           15          9   1.61
23        24    260       1500           14         13   1.84
24        25    420       2600           18          6   0.51
25        26     60        900            5         12   4.64
26        27    200       1700           11         12   2.72
27        28    280       1900           10          8   1.79
28        29    500       3000           15          4   0.48
29        30    460       2900           17          5   0.53

In [40]:
#Run a correlation on the dataset.
df.corr()


location     sales  employees  restaurants  foodcarts     price
location     1.000000  0.042705  -0.068923     0.049701   0.077219 -0.138444
sales        0.042705  1.000000   0.943238     0.913674  -0.919762 -0.966378
employees   -0.068923  0.943238   1.000000     0.856976  -0.874692 -0.881540
restaurants  0.049701  0.913674   0.856976     1.000000  -0.761793 -0.933951
foodcarts    0.077219 -0.919762  -0.874692    -0.761793   1.000000  0.860154
price       -0.138444 -0.966378  -0.881540    -0.933951   0.860154  1.000000

In [41]:
import statsmodels.formula.api as smf

In [42]:
#Regression
result = smf.ols('sales ~ employees + restaurants + foodcarts + price', data=df).fit()

In [43]:
result.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  sales   R-squared:                       0.981
Model:                            OLS   Adj. R-squared:                  0.977
Method:                 Least Squares   F-statistic:                     314.6
Date:                Thu, 07 Mar 2019   Prob (F-statistic):           5.52e-21
Time:                        15:57:11   Log-Likelihood:                -138.12
No. Observations:                  30   AIC:                             286.2
Df Residuals:                      25   BIC:                             293.2
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
===============================================================================
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
Intercept     319.6506     61.253      5.219      0.000     193.498     445.804
employees       0.0559      0.015      3.798      0.001       0.026       0.086
restaurants     3.1112      2.983      1.043      0.307      -3.032       9.254
foodcarts      -9.2350      2.695     -3.427      0.002     -14.786      -3.684
price         -47.9428     11.086     -4.325      0.000     -70.775     -25.111
==============================================================================
Omnibus:                        0.143   Durbin-Watson:                   2.490
Prob(Omnibus):                  0.931   Jarque-Bera (JB):                0.361
Skew:                          -0.012   Prob(JB):                        0.835
Kurtosis:                       2.463   Cond. No.                     2.81e+04
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 2.81e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

Results: Adj.R-squared is high (0.977) so model fits; intercept, foodcarts and price are significant IVs; p-factor significant. 

In [52]:
#Remove restaurants
result = smf.ols('sales ~ employees + foodcarts + price', data=df).fit()
result.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  sales   R-squared:                       0.980
Model:                            OLS   Adj. R-squared:                  0.977
Method:                 Least Squares   F-statistic:                     417.7
Date:                Thu, 07 Mar 2019   Prob (F-statistic):           4.19e-22
Time:                        16:32:06   Log-Likelihood:                -138.75
No. Observations:                  30   AIC:                             285.5
Df Residuals:                      26   BIC:                             291.1
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept    356.1201     50.379      7.069      0.000     252.565     459.675
employees      0.0615      0.014      4.476      0.000       0.033       0.090
foodcarts     -8.1718      2.499     -3.270      0.003     -13.309      -3.035
price        -56.9242      6.994     -8.139      0.000     -71.301     -42.548
==============================================================================
Omnibus:                        0.063   Durbin-Watson:                   2.619
Prob(Omnibus):                  0.969   Jarque-Bera (JB):                0.231
Skew:                           0.088   Prob(JB):                        0.891
Kurtosis:                       2.608   Cond. No.                     2.30e+04
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 2.3e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [ ]:
Adj-R-squared remains the same; p-factors significant.

In [53]:
#Run all variables without the intercept - set coefficient to 0
result = smf.ols(formula='sales ~ employees + restaurants + foodcarts + price - 1', data=df).fit()
result.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  sales   R-squared:                       0.987
Model:                            OLS   Adj. R-squared:                  0.986
Method:                 Least Squares   F-statistic:                     513.4
Date:                Thu, 07 Mar 2019   Prob (F-statistic):           2.53e-24
Time:                        16:34:13   Log-Likelihood:                -149.17
No. Observations:                  30   AIC:                             306.3
Df Residuals:                      26   BIC:                             311.9
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
===============================================================================
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
employees       0.0970      0.018      5.491      0.000       0.061       0.133
restaurants    11.9962      3.471      3.456      0.002       4.861      19.131
foodcarts      -5.6409      3.693     -1.527      0.139     -13.232       1.951
price          -9.0104     11.623     -0.775      0.445     -32.901      14.880
==============================================================================
Omnibus:                        0.713   Durbin-Watson:                   1.866
Prob(Omnibus):                  0.700   Jarque-Bera (JB):                0.706
Skew:                           0.074   Prob(JB):                        0.703
Kurtosis:                       2.263   Cond. No.                     3.98e+03
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 3.98e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [ ]:
Adj-R-squared improved (0.986); p-factor for price is high (0.445).

In [54]:
#Run above model without price 
result = smf.ols(formula='sales ~ employees + restaurants + foodcarts  - 1', data=df).fit()
result.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  sales   R-squared:                       0.987
Model:                            OLS   Adj. R-squared:                  0.986
Method:                 Least Squares   F-statistic:                     694.5
Date:                Thu, 07 Mar 2019   Prob (F-statistic):           1.18e-25
Time:                        16:37:02   Log-Likelihood:                -149.51
No. Observations:                  30   AIC:                             305.0
Df Residuals:                      27   BIC:                             309.2
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
===============================================================================
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
employees       0.0896      0.015      6.053      0.000       0.059       0.120
restaurants    13.7358      2.629      5.225      0.000       8.342      19.130
foodcarts      -8.4143      0.911     -9.240      0.000     -10.283      -6.546
==============================================================================
Omnibus:                        1.236   Durbin-Watson:                   1.798
Prob(Omnibus):                  0.539   Jarque-Bera (JB):                0.901
Skew:                           0.020   Prob(JB):                        0.637
Kurtosis:                       2.152   Cond. No.                         853.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [ ]:
Adj.R-squared remains the same; p-factors improve. 